network/data_link/at_parser.c

```diff
@@ -0,0 +1,1145 @@
+struct connection_desc {
+	/* Connection buffer */
+	volatile struct at_circular_buff	cbuff;
+	/* Pending bytes for reading */
+	uint32_t				to_read;
+	/* True if connection established */
+	bool					active;
+	/* Type of connection */
+	enum socket_type			type;
+};
+
+/* Structure storing the status of the parser */
+struct at_desc {
+	/* - Uart related fields */
+	/* Uart descriptor */
+	struct uart_desc	*uart_desc;
+	/* Irq descriptor */
+	struct irq_ctrl_desc	*irq_desc;
+	/* Uart irq id */
+	uint32_t		uart_irq_id;
+
+	//Connection related fields
+	/* Structures storing connections status */
+	struct connection_desc	conn[MAX_CONNECTIONS];
+	/* Id of active connection or -1 if non is active */
+	int32_t			current_conn;
+	/* Stores if running as single or multiple connection mode */
+	bool			multiple_conections;
+
+	/* - Data fields */
+	/* Static allocated buffers */
+	struct {
+		uint8_t	result_buff[RESULT_BUFF_LEN];
+		uint8_t	cmd_buff[CMD_BUFF_LEN];
+	} 			buffers;
+	/* Stores the output for the application */
+	struct circular_buffer	*cb_response;
+	/* Stores data received from the module */
+	volatile struct at_buff	result;
+	/* Buffer to build the command */
+	struct at_buff		cmd;
+	/* Buffer to read one char */
+	uint8_t			read_ch;
+
+	/* - Control fields */
+	/* State of the callback */
+	volatile enum {
+		/* Normal mode. Read each char and interpret the result */
+		READING_RESPONSES,
+		/* When an +IPD is received, callback enter in this mode */
+		READING_PAYLOAD,
+		/* Used when a reset command have been sent */
+		RESETTING_MODULE,
+		/* Used when using AT_SEND to wait for the character '>' */
+		WAITING_SEND
+
+	}			callback_operation;
+	/* Indexes in the async response given by the driver */
+	uint8_t			async_idx[NB_ASYNC_MESSAGES];
+	/* Indexes in the response given by the driver */
+	uint8_t			resp_idx[NB_RESPONSE_MESSAGES];
+	/* Ipd idx */
+	uint8_t			ipd_idx;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+static void at_sprintf(struct at_buff *dest, uint8_t *fmt, ...);
+static inline void at_sscanf(struct at_buff *src, uint8_t *fmt, uint32_t *p1,
+			     uint32_t *p2);
+static inline void at_strcpy(struct at_buff *dest, const struct at_buff *src);
+static inline void at_strcat(struct at_buff *dest, const struct at_buff *src);
+static inline void at_strcatch(struct at_buff *dest, uint8_t ch);
+static inline void at_set(struct at_buff *dest, uint8_t *buff, uint32_t len);
+
+/*
+ * How this function works it represented by this pseoudocode
+ * For each message and its index {
+ * 	if (ch == message[index]) {
+ * 		index++;
+ * 		if (index == strlen(message))
+ * 			return message_by_id
+ * 		}
+ * 	else
+ * 		index = 0;
+ *
+ * 	return -1;
+ * }
+ */
+static inline int32_t match_messages(const struct at_buff *messages,
```

Comment:
Should this method compare the messages to a string rather than a single character? Or do I understand it wrong?

Reply:
It only compares the new character received over uart, and if the character match at the saved index, the index is incremented.